src/counters.ts

```diff
@@ -0,0 +1,694 @@
+						};
+					}
+					if (!this.collection[processStatistics.pid][prop]) {
+						this.collection[processStatistics.pid][prop] = [];
+					}
+					trace(`prop: ${prop}, propColl: ${jsonDump(this.collection[processStatistics.pid][prop])}`);
+					this.collection[processStatistics.pid][prop].push(processStatistics[prop]);
+				});
+			});
+		};
+		// Kickoff the first collection
+		this.countersCollectionPromise = collectCounters();
+		// Set a timer for future collection(s)
+		this.countersTimer = setInterval(() => {
+			// start collecting counters unless we are already collecting one set.
+			if (!this.countersCollectionInProgress) {
+				this.countersCollectionInProgress = true;
+				this.countersCollectionPromise = collectCounters();
+				this.countersCollectionInProgress = false;
+			}
+
+		}, this.collectionInterval);
+	}
+
+
+	private async computeTotals(): Promise<void> {
+		const trace = require('debug')(`${logPrefix}:computeTotals:trace`);
+		// Totals are stored in a record corresponding to {@link Counters.TotalsProcessInfo}
+		trace(`collection: ${jsonDump(this.collection)}`);
+		trace(`collection.keys: ${jsonDump(Object.keys(this.collection))}`);
+		this.collection[Counters.TotalsProcessInfo.pid] = {
+			cpu: [],
+			memory: [],
+			ppid: Counters.TotalsProcessInfo.ppid,
+			pid: Counters.TotalsProcessInfo.pid,
+			ctime: [],
+			elapsed: this.collection[this.pid].elapsed, // elapsed time are same across all process so just refer to the one for the current process.
+			timestamp: this.collection[this.pid].timestamp, // timestamps  are same across all process so just refer to the one for the current process.
+		};
+		trace(`this.collection: ${jsonDump(this.collection)}`);
+		// compute and store totals for each timestamp value. 
+		for (let i in this.collection[Counters.TotalsProcessInfo.pid].timestamp) {
+			let cpu: number = 0;
+			let memory: number = 0;
+			let ctime: number = 0;
+			this.processesToTrack.forEach(proc => {
+				// we have collected anything for this process then add its stats to the totals
+				if (this.collection[proc.pid]) {
+					cpu += this.collection[proc.pid].cpu[i];
+					memory += this.collection[proc.pid].memory[i];
+					ctime += this.collection[proc.pid].ctime[i];
+				}
+			});
+			this.collection[Counters.TotalsProcessInfo.pid].cpu[i] = cpu;
+			this.collection[Counters.TotalsProcessInfo.pid].memory[i] = memory;
+			this.collection[Counters.TotalsProcessInfo.pid].ctime[i] = ctime;
+		}
+		trace(`this.collection[Counters.TotalsProcessInfo.pid]: ${jsonDump(this.collection[Counters.TotalsProcessInfo.pid])}`)
+	}
+
+	private async computeAndWriteStatistics(): Promise<void> {
+		const totalsStats: ProcessStatisticsCollection = this.collection[-1];
+		const datapoints: number = totalsStats.timestamp.length;
+		let p95: number;
+		let p90: number;
+		let p50: number;
+		[p50, p90, p95] = quantile(totalsStats.memory, [.5, .9, .95]);
+		const computedStats: ComputedStatistics = {

```

Comment:
Just as a note stuff like this is really where you should be adding comments. Without knowledge of statistics one would have no idea what this code is doing and what it's supposed to do.

Reply:
Will add comment.s

Reply:
Added comments on top of the ComputedStatistics interface definition to provide more context.